File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.py
from itertools import product


def calcul_discount_price(emoticons, discount_rate, index):
    return emoticons[index] * (100 - discount_rate) // 100


def check_service(users, emoticons, discount_rates):
    subscribers = 0
    total_price = 0

    for wanted_rate, price in users:
        sum_value = 0
        for index, discount_rate in enumerate(discount_rates):
            if wanted_rate <= discount_rate:
                sum_value += calcul_discount_price(emoticons, discount_rate, index)
            if sum_value >= price:
                subscribers += 1
                sum_value = 0
                break

        total_price += sum_value            

    return subscribers, total_price


def solution(users, emoticons):
    return max(check_service(users, emoticons, r) for r in product((10, 20, 30, 40), repeat=len(emoticons)))
        

File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.java
class Solution {

    private static final int[] DISCOUNT_OPTIONS = {10, 20, 30, 40};

    private static int[] emoticons;
    private static int[][] users;

    private static int maxSubscribers;
    private static int maxTotalPrice;

    public int[] solution(int[][] usersInput, int[] emoticonsInput) {
        users = usersInput;
        emoticons = emoticonsInput;
        maxSubscribers = 0;
        maxTotalPrice = 0;

        int[] discounts = new int[emoticons.length];
        dfs(0, discounts);

        return new int[]{maxSubscribers, maxTotalPrice};
    }

    private static void dfs(int depth, int[] discounts) {
        if (depth == emoticons.length) {
            evaluate(discounts);
            return;
        }

        for (int d : DISCOUNT_OPTIONS) {
            discounts[depth] = d;
            dfs(depth + 1, discounts);
        }
    }

    private static void evaluate(int[] discounts) {
        int subscribers = 0;
        int totalPrice = 0;

        for (int i = 0; i < users.length; i++) {
            int wantedRate = users[i][0];
            int limitPrice = users[i][1];
            int sum = 0;

            for (int j = 0; j < emoticons.length; j++) {
                if (discounts[j] >= wantedRate) {
                    sum += emoticons[j] * (100 - discounts[j]) / 100;
                }
                if (sum >= limitPrice) {
                    subscribers++;
                    sum = 0;
                    break;
                }
            }
            totalPrice += sum;
        }

        if (subscribers > maxSubscribers ||
            (subscribers == maxSubscribers && totalPrice > maxTotalPrice)) {
            maxSubscribers = subscribers;
            maxTotalPrice = totalPrice;
        }
    }
}